rustc_metadata: Optimize and document module children decoding

petrochenkov opened the pull request:
The first commit limits the item in the `item_children`/`each_child_of_item` query to modules (in name resolution sense) and adds a corresponding assertion.
The `associated_item_def_ids` query collecting children of traits and impls specifically now uses a simplified implementation not decoding unnecessary data instead of `each_child_of_item`, this gives a nice performance improvement.

The second commit does some renaming that clarifies the terminology used for all items in a module vs `use` items only.

petrochenkov committed:
rustc_metadata: Optimize and document module children decoding

rust-highfive commented:
Some changes occurred in src/tools/clippy.

cc @rust-lang/clippy

Some changes occurred in `clean/types.rs`.

cc @camelid

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

petrochenkov commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 62815a1f8ff0f1e06dd3f25b4aed88d32c2786f9 with merge 98c9bec5b9e4114d772623fd397e74d74b25da83...
<!-- homu: {"type":"TryBuildStarted","head_sha":"62815a1f8ff0f1e06dd3f25b4aed88d32c2786f9","merge_sha":"98c9bec5b9e4114d772623fd397e74d74b25da83"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4571063676?check_suite_focus=true)
Build commit: 98c9bec5b9e4114d772623fd397e74d74b25da83 (`98c9bec5b9e4114d772623fd397e74d74b25da83`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4571063676?check_suite_focus=true"},"merge_sha":"98c9bec5b9e4114d772623fd397e74d74b25da83"} -->

rust-timer commented:
Queued 98c9bec5b9e4114d772623fd397e74d74b25da83 with parent d3848cb659105d8384c00ba384ec2f6f39c47238, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d3848cb659105d8384c00ba384ec2f6f39c47238&end=98c9bec5b9e4114d772623fd397e74d74b25da83).

rust-timer commented:
Finished benchmarking commit (98c9bec5b9e4114d772623fd397e74d74b25da83): [comparison url](https://perf.rust-lang.org/compare.html?start=d3848cb659105d8384c00ba384ec2f6f39c47238&end=98c9bec5b9e4114d772623fd397e74d74b25da83).

**Summary**: This change led to very large relevant improvements 🎉 in compiler performance.
- Very large improvement in instruction counts (up to -14.5% on `incr-unchanged` builds of `issue-88862`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

petrochenkov committed:
rustc_middle: Rename `Export` to `ModChild` and add some comments

Also rename `module_exports`/`export_map` to `module_reexports`/`reexport_map` for clarity.

petrochenkov commented:
This should be ready now.

petrochenkov commented:
This is kind of a pre-requisite for https://github.com/rust-lang/rust/pull/91795, so it's better to land it sooner.
r? rust-lang/compiler

bors commented:
:umbrella: The latest upstream changes (presumably #92152) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jackh726 commented on compiler/rustc_metadata/src/rmeta/decoder.rs:
```diff
@@ -1071,15 +1071,30 @@ impl<'a, 'tcx> CrateMetadataRef<'a> {
+    /// including both proper items and reexports.
+    /// Module here is understood in name resolution sense - it can be a `mod` item,
+    /// or a crate root, or an enum, or a trait.
+    fn get_module_children(&self, id: DefIndex, mut callback: impl FnMut(Export), sess: &Session) {
```
I like the previous name more. Maybe something like `each_module_child`? `get*` feels like it should be returning something.

petrochenkov replied:
This is the only function in this module that uses a callback instead of returning something.
I want to change it to returning an iterator, it just needs to be simplified before that.

jackh726 commented on compiler/rustc_metadata/src/rmeta/decoder.rs:
```diff
@@ -1071,15 +1071,30 @@ impl<'a, 'tcx> CrateMetadataRef<'a> {
+
+    /// Iterates over all named children of the given module,
+    /// including both proper items and reexports.
+    /// Module here is understood in name resolution sense - it can be a `mod` item,
```
The overloading of the term `Module` is definitely confusing....

jackh726 reviewed (changes requested):
I think just one comment here.

For the most part looks *okay* to me, with the caveat that most of this code I've never really read before.

If it's not too much trouble, can you split the two major renames here to a separate commit? (`item_children` -> `module_children` and `each_child_of_item` -> `get_module_children`). That would make me more comfortable reviewing this,

petrochenkov committed:
rustc_metadata: Rename `item_children(_untracked)` to `module_children(_untracked)`

And `each_child_of_item` to `for_each_module_child`

petrochenkov commented:
Updated.

bors commented:
:umbrella: The latest upstream changes (presumably #92153) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #92244) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #92455) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #92034) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jackh726 commented:
@bors r+

bors commented:
:pushpin: Commit c3b644f7500c6559b44daa59bf0310cc829bb29b has been approved by `jackh726`

<!-- @bors r=jackh726 c3b644f7500c6559b44daa59bf0310cc829bb29b -->
<!-- homu: {"type":"Approved","sha":"c3b644f7500c6559b44daa59bf0310cc829bb29b","approver":"jackh726"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #92627) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
@bors r=jackh726

bors commented:
:pushpin: Commit 0e18e03e1872fe24db15aac6f56dcb9cdf1f044f has been approved by `jackh726`

<!-- @bors r=jackh726 0e18e03e1872fe24db15aac6f56dcb9cdf1f044f -->
<!-- homu: {"type":"Approved","sha":"0e18e03e1872fe24db15aac6f56dcb9cdf1f044f","approver":"jackh726"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #90639) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
@bors r=jackh726

bors commented:
:pushpin: Commit 4b03fd910c4d963a369baa3e38c9a7713c04a79c has been approved by `jackh726`

<!-- @bors r=jackh726 4b03fd910c4d963a369baa3e38c9a7713c04a79c -->
<!-- homu: {"type":"Approved","sha":"4b03fd910c4d963a369baa3e38c9a7713c04a79c","approver":"jackh726"} -->

bors commented:
:hourglass: Testing commit 4b03fd910c4d963a369baa3e38c9a7713c04a79c with merge d4135a2a8b08d4b9afb1ae4980ed1b8914b6af14...
<!-- homu: {"type":"BuildStarted","head_sha":"4b03fd910c4d963a369baa3e38c9a7713c04a79c","merge_sha":"d4135a2a8b08d4b9afb1ae4980ed1b8914b6af14"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

cjgillot commented:
@bors retry

bors commented:
:hourglass: Testing commit 4b03fd910c4d963a369baa3e38c9a7713c04a79c with merge e19ca1d946269f7b7eb13171531caf2e16f42076...
<!-- homu: {"type":"BuildStarted","head_sha":"4b03fd910c4d963a369baa3e38c9a7713c04a79c","merge_sha":"e19ca1d946269f7b7eb13171531caf2e16f42076"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/4752577254?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/d4135a2a8b08d4b9afb1ae4980ed1b8914b6af14/checks/4752577254/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4753154375?check_suite_focus=true)
Approved by: jackh726
Pushing e19ca1d946269f7b7eb13171531caf2e16f42076 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jackh726","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4753154375?check_suite_focus=true"},"merge_sha":"e19ca1d946269f7b7eb13171531caf2e16f42076"} -->

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#92086!

Tested on commit rust-lang/rust@e19ca1d946269f7b7eb13171531caf2e16f42076.
Direct link to PR: <https://github.com/rust-lang/rust/pull/92086>

💔 miri on windows: test-pass → build-fail (cc @eddyb @oli-obk @RalfJung).
💔 miri on linux: test-pass → build-fail (cc @eddyb @oli-obk @RalfJung).

rust-timer commented:
Finished benchmarking commit (e19ca1d946269f7b7eb13171531caf2e16f42076): [comparison url](https://perf.rust-lang.org/compare.html?start=02fe61b381c2dedc0071e1aacfbe91e0bad1f656&end=e19ca1d946269f7b7eb13171531caf2e16f42076).

**Summary**: This change led to very large relevant improvements 🎉 in compiler performance.
- Very large improvement in instruction counts (up to -14.9% on `incr-unchanged` builds of `issue-88862`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
